File: Lua/Common/ProtocolHelper.lua
local this = LuaObject:New()

ProtocolHelper = this

LoginType = {
	['Username'] = 'Username',
	['WeChat'] = 'WeChat',
	['Guest'] = 'Guest',
	['Phone'] = 'Phone',
}
this.account_login = "AccountService/account_login"
this.get_account = "AccountService/get_account"
--endregion 协议接口路径
this.OnSocketConnectTriggerAction = nil
this._LoginType = LoginType.Username;
-- this.NewLobbyLogin_lua = ''
-- this.NewGameLogin_lua = ''
this.entryRoom_password = ''

function this.Send_account_login()
	print('>>>>>>>>>>>>>>>>>>    send account_login')
	this._LoginType = LoginType.Username 
	local connectInfo = SocketConnectInfo.Instance;
	local bodyJson = {}
	bodyJson['username'] = connectInfo.lobbyUserName 
	bodyJson['password'] = connectInfo.lobbyPassword
	bodyJson['machineCode'] = "unity_" .. EginUser.Instance.device_id
	--bodyJson['device_id']  =  UnityEngine.SystemInfo.deviceUniqueIdentifier
	bodyJson['device_id']  =  "unity_" .. EginUser.Instance.device_id
	bodyJson['client_type'] = '0' 
	local loginJson = {}
	loginJson['type'] = 'AccountService'
	loginJson['tag'] = 'account_login'
	loginJson['body'] = bodyJson
	Util.SendPackage(cjson.encode(loginJson))
end

function this.Send_wechat_login()
	print('>>>>>>>>>>>>>>>>>>    send Send_wechat_login')
	if EginUser.Instance.wxOpenId == nil or EginUser.Instance.wxOpenId == "" then
		EginProgressHUD.Instance:ShowPromptHUD("登录数据有误,请稍后重新登录", 5.0);
		return;
	end
	if EginUser.Instance.wxUnionId == nil or EginUser.Instance.wxUnionId == "" then
		EginProgressHUD.Instance:ShowPromptHUD("登录数据有误,请稍后重新登录", 5.0);
		return;
	end
	this._LoginType = LoginType.WeChat 
	local bodyJson = {}
	bodyJson['openid'] = EginUser.Instance.wxOpenId
	bodyJson['unionid'] = EginUser.Instance.wxUnionId
	bodyJson['nickname'] = EginUser.Instance.wxNickname
	bodyJson['sex'] = EginUser.Instance.wxSex
	bodyJson['device_id']  =  "unity_" .. EginUser.Instance.device_id
	bodyJson['is_unity'] = 1
	bodyJson['agent_id'] = Utils.Agent_Id;
	local loginJson = {}
	loginJson['type'] = 'AccountService'
	loginJson['tag'] = 'wechat_oauth'
	loginJson['body'] = bodyJson

	UnityEngine.PlayerPrefs.SetString('WeChatUnionId',EginUser.Instance.wxUnionId)
	UnityEngine.PlayerPrefs.SetString('openid',EginUser.Instance.wxUnionId)
	UnityEngine.PlayerPrefs.SetString('nickname',EginUser.Instance.wxNickname)
	UnityEngine.PlayerPrefs.SetString('device_id',"unity_" .. EginUser.Instance.device_id)
	
	UnityEngine.PlayerPrefs.Save()
	Util.SendPackage(cjson.encode(loginJson))
end
function this.Send_wechat_login_New()
	print('>>>>>>>>>>>>>>>>>> NewNNNNNNNNN Send_wechat_login')
	this._LoginType = LoginType.WeChat 
	local bodyJson = {}
	bodyJson['code'] = ''
	bodyJson['state'] = ''
	bodyJson['device_id']  =  "unity_" .. EginUser.Instance.device_id
	-- bodyJson['openid'] = EginUser.Instance.wxOpenId
	-- bodyJson['nickname'] = EginUser.Instance.wxNickname
	-- bodyJson['sex'] = EginUser.Instance.wxSex
	-- bodyJson['is_unity'] = 1
	local loginJson = {}
	loginJson['type'] = 'AccountService'
	loginJson['tag'] = 'wechat_oauth1'
	loginJson['body'] = bodyJson
	Util.SendPackage(cjson.encode(loginJson))
end

function this.Send_guest_login(  )
	print('>>>>>>>>>>>>>>>>>>     Send_guest_login')
	this._LoginType = LoginType.Guest 
	local bodyJson = {}
	local uid_long =0 
	local GuestUID = UnityEngine.PlayerPrefs.GetString("GuestUID",tostring(uid_long));
	if GuestUID =='0' then
		bodyJson['uid'] = uid_long
	else
		bodyJson['uid'] = GuestUID
	end
	bodyJson['device_id']  =  "unity_" .. EginUser.Instance.device_id
	bodyJson['agent_id'] = Utils.Agent_Id;
	local loginJson = {}
	loginJson['type'] = 'AccountService'
	loginJson['tag'] = 'guest_login'
	loginJson['body'] = bodyJson

	Util.SendPackage(cjson.encode(loginJson))

end

function this.Send_get_account(  )
	--error('>>>>>>>>>>>>>>>>>>   Send_get_account')

	local loginJson = {}
	loginJson['type'] = 'AccountService'
	loginJson['tag'] = 'get_account'
	loginJson['body'] = ''
	Util.SendPackage(cjson.encode(loginJson))
end

function this.Receive_get_ccount(result )
	print('>>>>>>>>>>>>>>>>>>   Receive_get_ccount')
 	local  session = "";
    --对socket协议中 字段 的不同进行修正
    -- local resultDict = cjson.decode(tostring(result))--cjson.decode(result)
    -- print(result)
    -- print(resultDict['userid'])
    -- print(resultDict['avatar_no'])
    
    -- if resultDict["id"] ~= nil then resultDict["id"] = resultDict["userid"] end
    -- if resultDict["exp"] ~= nil then resultDict["exp"] = resultDict["exp_value"] end
    -- if resultDict["next_exp"] ~= nil then resultDict["next_exp"] = resultDict["exp_value"] end
    -- if resultDict["avatar_img"] ~= nil then resultDict["avatar_img"] = "" end
    -- if resultDict["agent"] ~= nil then resultDict["agent"] = "" end
    -- if resultDict["weak"] ~= nil then resultDict["weak"] = "0" end
    -- if resultDict["gold_money"] ~= nil then resultDict["gold_money"] = "0" end
    --处理登录返回信息
    EginUser.Instance:InitUserWithDict(result, session);
    EginUser.Instance.isGuest = false
end

function this.Send_game_login(  )
	print('>>>>>>>>>>>>>>>>>>   Send_game_login')

	local tConnectInfo = SocketConnectInfo.Instance
	local tMessageBody = {}
	tMessageBody['userid'] = tConnectInfo.userId
	tMessageBody['password'] = tConnectInfo.userPassword
	tMessageBody['dbname'] = tConnectInfo.roomDBName
	tMessageBody['version'] = '0'
	tMessageBody['client_type'] = '0'
	tMessageBody['roomid'] = tConnectInfo.roomId
	tMessageBody['fixseat'] = true
	if(PlatformGameDefine.game.GameID == "1006" and PlatformGameDefine.game.GameTypeIDs == "20")then
		if string.len(this.entryRoom_password) >0 then
			tMessageBody['entry_pwd'] = this.entryRoom_password
		else
			tMessageBody['client_info'] = 'WIN'
		end
	else
		tMessageBody['client_info'] = 'WIN';
	end
	
	local tMessageObj  = {}
	tMessageObj['type'] = 'account'
	tMessageObj['tag'] = 'login'
	tMessageObj['body'] = tMessageBody 
	local tMsg =cjson.encode(tMessageObj)
	Util.SendPackageSpecial(tMsg,tConnectInfo.userPassword);
	--this.entryRoom_password = ""
end


function this.Send_Socketlobby_login()
	print('>>>>>>>>>>>>>>>>>>   Send_Socketlobby_login')
	if this.OnSocketConnectTriggerAction ~= nil then
		this.OnSocketConnectTriggerAction()
		this.OnSocketConnectTriggerAction = nil 
		return
	end 
	if this._LoginType  == LoginType.WeChat then
		this.Send_wechat_login()
	elseif this._LoginType  ==LoginType.Guest then
		this.Send_guest_login()
	elseif this._LoginType == LoginType.Username then
		this.Send_account_login()
	elseif this._LoginType == LoginType.Phone then

		this.Send_Phone_login()

	end
end

function this.Send_login(pIsGame)
	-- print(">>>>>>>>>>>>>>>  send Login ")
	-- print(  pIsGame  )
	if pIsGame == nil then
		pIsGame = false 
	end
	if pIsGame then
		this.Send_game_login()
	else
		this.Send_Socketlobby_login()
	end

end

function this.Send( pPath,pBodyJson )
	local tSendJson = {}
	local tStr =string.split(pPath,"/");
	if tStr == nil or #tStr ~=2  then
		error("the format of path \"" .. pPath .. "\" is wrong")
	end
	tSendJson['type'] = tStr[1]
	tSendJson['tag'] = tStr[2]
	tSendJson['body'] = pBodyJson
	Util.SendPackage(tSendJson)
end

----添加手机登录 --  
function this.Send_Phone_login()
	print('>>>>>>>>>>>>>>>>>>   Send_Phone_login')

 	this._LoginType =  LoginType.Phone 
	local bodyJson = {}
	bodyJson['phone'] = EginUser.Instance.phoneNum
	bodyJson['device_id']  =  "unity_" .. EginUser.Instance.device_id
	bodyJson['phonecode'] = EginUser.Instance.phonecode 
	bodyJson['agent_id'] = Utils.Agent_Id;

	local loginJson = {}
	loginJson['type'] = 'AccountService'
	loginJson['tag'] = 'mobile_oauth'
	loginJson['body'] = bodyJson
	Util.SendPackage(cjson.encode(loginJson))
 end


--获取手机验证码
function this.Send_Phone_IdentifyCode()
	print('>>>>>>>>>>>>>>>>>>   Send_Phone_IdentifyCode')
	local tBody = {}
	tBody['phone'] = EginUser.Instance.phoneNum
	local loginJson = {}
	loginJson['type'] = 'AccountService'
	loginJson['tag'] = 'send_mobile_sms'
	loginJson['body'] = tBody
	Util.SendPackage(cjson.encode(loginJson))
end

File: ___HappyCityScripts/Helper/ProtocolHelper.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class ProtocolHelper {

    #region 协议接口路径. 这部分只是字符串路径,如果有热更新需要可以在Lua中直接封装
    public const string account_login = "AccountService/account_login";
    public const string get_account = "AccountService/get_account";
    //public const string account_login = "AccountService/account_login";
    //public const string account_login = "AccountService/account_login";
    //public const string account_login = "AccountService/account_login";
    #endregion 协议接口路径
        
    public static System.Action OnSocketConnectTriggerAction;
    public static LoginType _LoginType = LoginType.Username;
    public static string NewLobbyLogin_lua = string.Empty;
    public static string NewGameLogin_lua = string.Empty;

    #region lobby大厅 socket相关协议
    /// <summary>
    /// socket登录协议
    /// 请求Send_login自动判断登入lobby还是登入game
    /// </summary>
    /// <param name="username"></param>
    /// <param name="password"></param>
    public static void Send_account_login()
    {
       
        _LoginType = LoginType.Username;
		SocketConnectInfo connectInfo = SocketConnectInfo.Instance;
        JSONObject bodyJson = new JSONObject();
        Debug.Log("====SendAccount  Login================    "+ connectInfo.lobbyUserName);
        bodyJson.AddField("username", connectInfo.lobbyUserName);
        bodyJson.AddField("password", connectInfo.lobbyPassword);
        //bodyJson.AddField("captcha", kPassword.value);
        //bodyJson.AddField("proving", kPassword.value);
//#if UNITY_EDITOR
//        bodyJson.AddField("machineCode", "123456");//测试用
//        bodyJson.AddField("device_id", "123456");//测试用
//#else
        bodyJson.AddField("machineCode", "unity_" + EginUser.Instance.device_id);
        bodyJson.AddField("device_id", "");//"unity_" + EginUser.Instance.device_id);//测试用
//#endif
        bodyJson.AddField("client_type", "0");

        JSONObject loginJson = new JSONObject();
        loginJson.AddField("type", "AccountService");
        loginJson.AddField("tag", "account_login");
        loginJson.AddField("body", bodyJson);

		BaseSceneLua.SocketSendMessage(loginJson);
    }

    /// <summary>
    /// 发送socket微信登录协议
    /// </summary>
    public static void Send_wechat_login()
    {
        ProtocolHelper._LoginType = LoginType.WeChat;
        JSONObject json = new JSONObject();
        json.AddField("openid", EginUser.Instance.wxOpenId);
        json.AddField("nickname", EginUser.Instance.wxNickname);
        json.AddField("sex", EginUser.Instance.wxSex);
        json.AddField("is_unity", 1);
        JSONObject loginJson = new JSONObject();
        loginJson.AddField("type", "AccountService");
        loginJson.AddField("tag", "wechat_oauth");
        loginJson.AddField("body", json);

		BaseSceneLua.SocketSendMessage(loginJson);
    }

    public static void Send_guest_login() {
        _LoginType = LoginType.Guest;
        JSONObject json = new JSONObject();
        long uid_long = 0;
        string GuestUID = PlayerPrefs.GetString("GuestUID", "" + uid_long);
        //UnityEngine.Debug.Log("CK : ----------------GuestUIDGuestUIDGuestUID-------------- GuestUID = " + GuestUID);
        
        if(!long.TryParse(GuestUID, out uid_long))
        {
            GuestUID = ""+uid_long;
        }

        if(uid_long == 0) json.AddField("uid", uid_long);//0 的时候只能用数字
        else json.AddField("uid", GuestUID);
        JSONObject loginJson = new JSONObject();
        loginJson.AddField("type", "AccountService");
        loginJson.AddField("tag", "guest_login");
        loginJson.AddField("body", json);

		BaseSceneLua.SocketSendMessage(loginJson);
    }

    /// <summary>
    /// 发送
    /// socket登入成功后, 获取用户信息
    /// </summary>
    public static void Send_get_account()
    {
        JSONObject json = new JSONObject();
        json.AddField("type", "AccountService");
        json.AddField("tag", "get_account");
        json.AddField("body", JSONObject.nullJO);
        BaseSceneLua.SocketSendMessage(json);
    }

    /// <summary>
    /// 接收
    /// 获取用户信息, 返回数据的处理
    /// </summary>
    /// <param name="result"></param>
    public static void Receive_get_ccount(JSONObject result)
    {
        string session = "";
        //if (www.responseHeaders.ContainsKey("SET-COOKIE"))
        //{
        //    session = www.responseHeaders["SET-COOKIE"];
        //}
        //Dictionary<string, string> resultDict = ((JSONObject)result.resultObject).ToDictionary();
        //对socket协议中 字段 的不同进行修正
        Dictionary<string, string> resultDict = result.ToDictionary();
        if (!resultDict.ContainsKey("id")) resultDict["id"] = resultDict["userid"];
        if (!resultDict.ContainsKey("exp")) resultDict["exp"] = resultDict["exp_value"];
        if (!resultDict.ContainsKey("next_exp")) resultDict["next_exp"] = resultDict["exp_value"];
        if (!resultDict.ContainsKey("avatar_img")) resultDict["avatar_img"] = "";
        if (!resultDict.ContainsKey("agent")) resultDict["agent"] = "";
        if (!resultDict.ContainsKey("weak")) resultDict["weak"] = "0";
        if (!resultDict.ContainsKey("gold_money")) resultDict["gold_money"] = "0";

        //处理登录返回信息
        EginUser.Instance.InitUserWithDict(resultDict, session);
        EginUser.Instance.isGuest = false;
    }
#endregion lobby大厅 socket相关协议

#region 游戏 socket相关协议
	//斗地主电视现场版房间在进入房间时需要密码.
	public static string entryRoom_password = "";
    /// <summary>
    /// 游戏登入协议
    /// 请求Send_login自动判断登入lobby还是登入game
    /// </summary>
    public static void Send_game_login()
    {
        if (!string.IsNullOrEmpty(NewGameLogin_lua))
        {
            SimpleFramework.Util.CallMethod("ProtocolHelper", NewGameLogin_lua);
            return;
        }

        SocketConnectInfo connectInfo = SocketConnectInfo.Instance;
        JSONObject messageBody = new JSONObject();
        messageBody.AddField("userid", connectInfo.userId);
        messageBody.AddField("password", connectInfo.userPassword);
        messageBody.AddField("dbname", connectInfo.roomDBName);
        messageBody.AddField("version", "0");                   ///0
        messageBody.AddField("client_type", "0");
        messageBody.AddField("roomid", connectInfo.roomId);
        messageBody.AddField("fixseat", true);
		//斗地主电视现场版房间在进入房间时需要密码.
		if(entryRoom_password.Length > 0){
			messageBody.AddField("entry_pwd", entryRoom_password);
		}else{
			//进入密码房间发client_info服务器会收不到消息，奇怪
			messageBody.AddField("client_info", "WIN");          // 0928  UNITY  解决排行榜 不显示的问题  更改  此处 .
		}
        //messageBody.AddField("first_money", "");
        JSONObject messageObj = new JSONObject();
        messageObj.AddField("type", "account");
        messageObj.AddField("tag", "login");
        messageObj.AddField("body", messageBody);
        SocketManager.Instance.SendPackageWithJson(messageObj);
		entryRoom_password = "";
    }
#endregion 游戏 socket相关协议

#region 辅助方法,helper
    public static void Send_Socketlobby_login()
    {
        UnityEngine.Debug.Log("CK : ------------]]]]]]]]]]]]]]]]]]]]]------------------ _LoginType = " + _LoginType);

        if(OnSocketConnectTriggerAction != null)
        {
            OnSocketConnectTriggerAction();
            OnSocketConnectTriggerAction = null;
            return;
        }

        if (!string.IsNullOrEmpty(NewLobbyLogin_lua))
        {
            SimpleFramework.Util.CallMethod("ProtocolHelper", NewLobbyLogin_lua);
            return;
        }

        switch (_LoginType)
        {
            case LoginType.WeChat:
                Send_wechat_login();
                break;
            case LoginType.Guest:
                Send_guest_login();
                break;
            default:
                Send_account_login();
                break;
        }
    }

    public static void Send_login()
    {
        Send_login(false);
    }

    public static void Send_login(bool isGame)
    {

        Debug.Log("))))))))))))))))))))  send  login "+ isGame.ToString());

        SimpleFramework.Util.CallMethod("ProtocolHelper", "Send_login",isGame);
        return;
        if (isGame)
        {
            //NewLobbyLogin_lua = "Send_game_login";
            //if (!string.IsNullOrEmpty(NewLobbyLogin_lua))
            //{
            //    SimpleFramework.Util.CallMethod("ProtocolHelper", NewLobbyLogin_lua);
            //    NewLobbyLogin_lua = "";
            //    return;
            //}
           
            Send_game_login();
        }
        else
        {
            //NewLobbyLogin_lua = "Send_Socketlobby_login";
            //if (!string.IsNullOrEmpty(NewLobbyLogin_lua))
            //{
            //    SimpleFramework.Util.CallMethod("ProtocolHelper", NewLobbyLogin_lua);
            //    NewLobbyLogin_lua = "";
            //    return;
            //}
            Send_Socketlobby_login();
        }
    }

    public static void Send(string path, JSONObject bodyJson)
    {
        JSONObject sendJson = new JSONObject();
        string[] split = path.Split('/');
        if (split == null || split.Length != 2) throw new System.Exception("the format of path \"" + path + "\" is wrong");
        sendJson.AddField("type", split[0]);
        sendJson.AddField("tag", split[1]);
        sendJson.AddField("body", bodyJson);
		BaseSceneLua.SocketSendMessage(sendJson);
    }
#endregion 辅助方法
}

public enum LoginType{
    Username = 0,
    WeChat,
    Guest,
}
